### PR TITLE
feature: add before/after InvocationRequest hooks

### PR DESCRIPTION
**Pull request by markwolff**
Copied from #338 

Adds two `InvocationRequest` hooks to `WorkerChannel` to enable context correlation and node.js worker APM instrumentation. Sample use case for Azure Monitor node.js SDK.

- `invocationRequestBefore`: `(context, userFunction) => contextWrappedUserFunction`
- `invocationRequestAfter`: `() => void`

> note: `workerChannel` would need to be passed to app insights in some way, i.e. in `nodejsWorker.ts`:
```ts
if (shouldSetupAppInsightsFunctionsIntegration) {
  require("/path/to/applicationinsights/agent/functions")
    .setupFunctionsInvocationHooks(workerChannel);
}
```

In App Insights SDK, `setupFunctionsInvocationHooks` would be defined:
```ts
workerChannel.invocationRequestBefore = (context, userFunction) => {
    // Map Functions context to AppInsights context
    const correlationContext = appInsights.startOperation(context, "Functions.JavaScript.Trigger");

    // Wrap the Function runtime with correlationContext
    return appInsights.wrapWithCorrelationContext(userFunction, correlationContext);
}

workerChannel.invocationRequestAfter = () => {
    // Do any extra cleanup or flushing here
    appInsights.defaultClient.flush(); // can be async or sync
}
```

### CODE REVIEW

File: src/WorkerChannel.ts [hunk 1]
```diff
@@ -160,6 +178,8 @@ export class WorkerChannel implements IWorkerChannel {
         invocationId: msg.invocationId,
         result: this.getStatus(err)
       }
+      
+      this.runInvocationRequestAfter(context);
```
Comment (mhoeger) [inline src/WorkerChannel.ts]: Remove this?

File: test/WorkerChannelTests.ts [hunk 2]
```diff
@@ -226,67 +283,114 @@ describe('WorkerChannel', () => {
...
+  });
+
+  it ('invokes function', () => {
+    loader.getFunc.returns((context) => context.done());
+    loader.getInfo.returns({
+      name: 'test',
+      outputBindings: {}
+    })
+
+    assertInvokedFunction();
```
Comment (mhoeger) [inline test/WorkerChannelTests.ts]: thanks for tests :)

File: src/WorkerChannel.ts [hunk 3]
```diff
@@ -304,4 +333,18 @@ export class WorkerChannel implements IWorkerChannel {
 
     return status;
   }
+
+  private runInvocationRequestBefore(context: Context, userFunction: Function): Function {
+    let wrappedFunction = userFunction;
+    for (let before of this._invocationRequestBefore) {
+      wrappedFunction = before(context, wrappedFunction);
```
Comment (mhoeger) [inline src/WorkerChannel.ts:344]: Just double checking that you don't want to make anything async? Could do a Promise.all

File: src/WorkerChannel.ts (hunk 1 above)
Comment (markwolff) [reply src/WorkerChannel.ts]: I can move this to be the `else` case for the `.then` Promise checks below if you think that makes sense. Is there a better place for this to live to handle UserFunctions that have callbacks?

File: src/WorkerChannel.ts (hunk 3 above)
Comment (markwolff) [reply src/WorkerChannel.ts:344]: I could see these being useful as Promises. Currently each `before` updates `wrappedFunction` on completion. Is there a nice way to accomplish this with Promise.all?

File: src/WorkerChannel.ts (hunk 1 above)
Comment (mhoeger) [reply src/WorkerChannel.ts]: Ahh sorry - pointed to the wrong parts to remove!! Can you remove the calls in line 221 and 225 to "runInvocationRequestAfter"? In lines 222 and 226, "context.done" is called, which is the same as this "resultCallback". So in cases of async / promise-returning user functions, "runInvocationRequestAfter" would be called twice

Comment (markwolff) [reply src/WorkerChannel.ts]: Good catch! Refactored the tests a bit to support tests for promise user functions

File: src/WorkerChannel.ts (hunk 3 above)
Comment (mhoeger) [reply src/WorkerChannel.ts:344]: Thoughts on await on `before` and `after` calls? Otherwise this PR looks good!

**Review by mhoeger — APPROVED** [review]
✅ LGTM - on registering async functions, since it's not like a one line change, let's just cross that bridge when we get there.

Thanks!!

**Pull request merged**